File: ats-backend/utils/parser.js
const fs = require('fs');
const path = require('path');
const pdfParse = require('pdf-parse');
const natural = require('natural');

// List of predefined job-related skill phrases
const skill_phrases = [
  'JavaScript', 'React', 'Node.js', 'MongoDB', 'Express.js',
  'Python', 'Java', 'HTML', 'CSS', 'C++', 'Machine Learning',
  'Deep Learning', 'Data Science', 'SQL', 'NoSQL', 'Docker',
  'Kubernetes', 'AWS', 'Azure', 'Git', 'REST APIs', 'Linux'
];

// Extract text from resume buffer
async function extractTextFromResume(buffer) {
  const data = await pdfParse(buffer);
  return data.text;
}

// Extract matched skills from resume text
function extract_skills(text) {
  const tokens = text.toLowerCase().split(/[^a-zA-Z0-9+.#]/);
  const matched = [];

  for (const phrase of skill_phrases) {
    const lowerPhrase = phrase.toLowerCase();
    const phraseTokens = lowerPhrase.split(' ');

    if (phraseTokens.length === 1) {
      if (tokens.includes(lowerPhrase)) matched.push(phrase);
    } else {
      if (text.toLowerCase().includes(lowerPhrase)) matched.push(phrase);
    }
  }

  return [...new Set(matched)];
}

// Score = % of job skills matched
function calculate_manual_score(resumeSkills, jobSkills) {
  if (!jobSkills || jobSkills.length === 0) return { score: 0, matched: [] };

  const matched = resumeSkills.filter(skill =>
    jobSkills.map(s => s.toLowerCase()).includes(skill.toLowerCase())
  );

  const score = Math.round((matched.length / jobSkills.length) * 100);
  return { score, matched };
}

module.exports = {
  extractTextFromResume,
  extract_skills,
  calculate_manual_score,
};

File: ats-backend/resume-parser/parser.py
import pytesseract
from pdf2image import convert_from_path
import pdfplumber
import nltk
import spacy
import re
import json
from nltk.corpus import stopwords
from nltk.tokenize import TreebankWordTokenizer
from PIL import Image, ImageEnhance, ImageFilter
from difflib import get_close_matches
import sys

# Setup
nltk.download('stopwords')
nltk.download('punkt')

stop_words = set(stopwords.words('english'))
nlp = spacy.load('en_core_web_sm')
pytesseract.pytesseract.tesseract_cmd = r"C:\Program Files\Tesseract-OCR\tesseract.exe"

COMMON_SKILLS = [
    "python", "java", "c++", "sql", "react", "node", "html", "css", "javascript",
    "aws", "docker", "kubernetes", "machine learning", "data science", "mongodb",
    "flask", "django", "excel", "power bi", "tableau", "git", "github", "pandas",
    "numpy", "communication", "leadership", "teamwork", "problem solving"
]

def extract_text_from_pdf(file_path):
    with pdfplumber.open(file_path) as pdf:
        text = "\n".join(page.extract_text() or "" for page in pdf.pages)
    return text.strip()

def extract_text_from_pdf_ocr(file_path):
    pages = convert_from_path(file_path, dpi=300)
    full_text = ''
    for page in pages:
        image = page.convert('L')
        image = image.filter(ImageFilter.SHARPEN)
        enhancer = ImageEnhance.Contrast(image)
        image = enhancer.enhance(2)
        text = pytesseract.image_to_string(image)
        full_text += text + '\n'
    return full_text.strip()

def extract_email(text):
    text = text.replace(' ', '')
    emails = re.findall(r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}", text)
    return emails[0] if emails else ""

def extract_phone(text):
    match = re.search(r"(\+91[\s-]?)?[0]?[789]\d{9}", text)
    return match.group() if match else ""

def extract_skills(text):
    text = text.lower()
    text = re.sub(r'[^\w\s]', ' ', text)
    tokenizer = TreebankWordTokenizer()
    tokens = tokenizer.tokenize(text)
    tokens = [t for t in tokens if t.isalpha() and t not in stop_words]
    
    matched_skills = set()
    for skill in COMMON_SKILLS:
        skill_tokens = skill.lower().split()
        if all(token in tokens for token in skill_tokens):
            matched_skills.add(skill)
        else:
            close = get_close_matches(skill, tokens, n=1, cutoff=0.85)
            if close:
                matched_skills.add(skill)

    return list(matched_skills)

def compute_score(matched_skills, job_description):
    job_desc = job_description.lower()
    job_desc = re.sub(r'[^\w\s]', ' ', job_desc)
    tokenizer = TreebankWordTokenizer()
    job_tokens = tokenizer.tokenize(job_desc)
    job_tokens = [t for t in job_tokens if t.isalpha()]

    jd_skill_matches = set()
    for skill in COMMON_SKILLS:
        skill_tokens = skill.split()
        if all(word in job_tokens for word in skill_tokens):
            jd_skill_matches.add(skill)

    if not jd_skill_matches:
        return 0.0

    intersection = set(matched_skills) & jd_skill_matches
    return round(len(intersection) / len(jd_skill_matches) * 100, 2)

def main(file_path, job_description):
    text = extract_text_from_pdf(file_path)
    if not text or len(text.strip()) < 100:
        text = extract_text_from_pdf_ocr(file_path)

    email = extract_email(text)
    phone = extract_phone(text)
    skills = extract_skills(text)
    score = compute_score(skills, job_description)

    result = {
        "email": email,
        "phone": phone,
        "skills": skills,
        "score": score
    }

    print(json.dumps(result))  # ✅ Only JSON

if __name__ == "__main__":
    if len(sys.argv) != 3:
        pass  # No print to avoid breaking JSON
    else:
        main(sys.argv[1], sys.argv[2])
